js/gateio.js

```diff
@@ -2035,23 +2070,66 @@ module.exports = class gateio extends Exchange {
             //     }
             // }
         }
-        if (type === 'limit') {
-            if (!price) {
-                throw new ArgumentsRequired ('Argument price is required for ' + this.id + '.createOrder for limit orders');
-            }
-            request['price'] = this.priceToPrecision (symbol, price);
-        } else if (type === 'market' && contract) {
+        if (type === 'market' && market['contract']) {
             request['tif'] = 'ioc';
             request['price'] = 0;
+        } else {
+            request['price'] = price;
         }
-        const method = this.getSupportedMapping (market['type'], {
-            'spot': 'privateSpotPostOrders',
-            // 'margin': 'privateSpotPostOrders',
-            'swap': 'privateFuturesPostSettleOrders',
-            'future': 'privateDeliveryPostSettleOrders',
-        });
-        const response = await this[method] (this.extend (request, params));
-        return this.parseOrder (response, market);
+        return this.extend (request, params);
+    }
+
+    stopLossRequest (symbol, type, side, amount, price = undefined, params = {}) {
+        const market = this.market (symbol);
+        const uppercaseType = type.toUpperCase ();
+        const stop_limit = uppercaseType === 'STOP_LOSS_LIMIT' || uppercaseType === 'TAKE_PROFIT_LIMIT';
+        const stopPrice = this.safeNumber (params, 'stopPrice');
+        if (stopPrice === undefined) {
+            throw new InvalidOrder (this.id + ' createOrder() requires a stopPrice extra param for a ' + type + ' order');
+        }
+        params = this.omit (params, 'stopPrice');
+        const request = {};
+        const trigger = {
+            'price': this.priceToPrecision (symbol, stopPrice),
+        };
+        const expiration = this.safeValue (params, 'expiration');
+        if (expiration) {
+            // How long (in seconds) to wait for the condition to be triggered before cancelling the order.
+            trigger['expiration'] = expiration;
+        } else if (!market['contract']) {
+            throw new InvalidOrder (this.id + ' createOrder() requires an expiration extra param for a ' + type + ' order');
+        }
+        const defaultTif = stop_limit ? 'gtc' : 'ioc';
+        if (market['contract']) {
+            trigger['rule'] = side === 'buy' ? 1 : 2;
+            trigger['strategy_type'] = this.safeValue (params, 'strategy_type', 0);
+            trigger['price_type'] = this.safeValue (params, 'price_type');
+            const initial = this.safeValue (params, 'initial');
+            request['initial'] = this.extend ({
```

Comment:
extend is used so that orders like this will still work

```
exchange
  .createOrder("BTC/USDT", "stop_loss_limit", "sell", 0.0001, 60000, {
    stopPrice: 62000,
    expiration: 3600,
    trigger: {
      price: "60000",
      rule: "<=",
      expiration: 3600,
    },
    put: {
      type: "limit",
      side: "sell",
      price: "62000",
      amount: "0.0001",
      account: "normal",
      time_in_force: "gtc",
    },
    market: "BTC_USDT",
  })
```